presto-main/src/main/java/io/prestosql/cost/CostCalculatorUsingExchanges.java

```diff
@@ -234,67 +247,127 @@ public PlanNodeLocalCostEstimate visitExchange(ExchangeNode node, Void context)
                     node.getDistributionType() == SpatialJoinNode.DistributionType.REPLICATED);
+            return costForLookupJoin(node, localCost);
         }
 
         @Override
-        public PlanNodeLocalCostEstimate visitValues(ValuesNode node, Void context)
+        public PlanCostEstimate visitValues(ValuesNode node, Void context)
         {
-            return PlanNodeLocalCostEstimate.zero();
+            return costForSource(node, PlanNodeLocalCostEstimate.zero());
         }
 
         @Override
-        public PlanNodeLocalCostEstimate visitEnforceSingleRow(EnforceSingleRowNode node, Void context)
+        public PlanCostEstimate visitEnforceSingleRow(EnforceSingleRowNode node, Void context)
         {
-            return PlanNodeLocalCostEstimate.zero();
+            return costForAccumulation(node, PlanNodeLocalCostEstimate.zero());
         }
 
         @Override
-        public PlanNodeLocalCostEstimate visitLimit(LimitNode node, Void context)
+        public PlanCostEstimate visitLimit(LimitNode node, Void context)
         {
             // This is just a wild guess. First of all, LimitNode is rather rare except as a top node of a query plan,
             // so proper cost estimation is not that important. Second, since LimitNode can lead to incomplete evaluation
             // of the source, true cost estimation should be implemented as a "constraint" enforced on a sub-tree and
             // evaluated in context of actual source node type (and their sources).
-            return PlanNodeLocalCostEstimate.ofCpu(getStats(node).getOutputSizeInBytes(node.getOutputSymbols(), types));
+            PlanNodeLocalCostEstimate localCost = PlanNodeLocalCostEstimate.ofCpu(getStats(node).getOutputSizeInBytes(node.getOutputSymbols(), types));
+            return costForStreaming(node, localCost);
         }
 
         @Override
-        public PlanNodeLocalCostEstimate visitUnion(UnionNode node, Void context)
+        public PlanCostEstimate visitUnion(UnionNode node, Void context)
         {
-            // Cost will be accounted either in CostCalculatorUsingExchanges#CostEstimator#visitExchanged
+            // Cost will be accounted either in CostCalculatorUsingExchanges#CostEstimator#visitExchange
             // or in CostCalculatorWithEstimatedExchanges#CostEstimator#visitUnion
             // This stub is needed just to avoid the cumulative cost being set to unknown
-            return PlanNodeLocalCostEstimate.zero();
+            return costForStreaming(node, PlanNodeLocalCostEstimate.zero());
+        }
+
+        private PlanCostEstimate costForSource(PlanNode node, PlanNodeLocalCostEstimate localCost)
+        {
+            verify(node.getSources().isEmpty(), "Unexpected sources for %s: %s", node, node.getSources());
+            return localCost.toPlanCost();
+        }
+
+        private PlanCostEstimate costForAccumulation(PlanNode node, PlanNodeLocalCostEstimate localCost)
+        {
```

Comment:
`verify(sourcesCosts.size() == 1);` like in `costForLookupJoin `

Reply:
`costForLookupJoin` is different, as it treats sources non-uniformly. If it didn't check there are exactly 2 sources, some sources could remain ignored in costs calculations.